<div style="color:#FF0000         ; display  :fill; border-radius:90px;
           background-color:#FFFF00; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :##FF0000; 
          text-align: Left;   font-size:28px; font-family  :cursive">       
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;Importing the Libraries<a id=5></a></p>
</div>

In [1]:
# import classes for Data Analysis
import numpy  as np
import pandas as pd
from fastai.tabular.all import *

import warnings
warnings.filterwarnings("ignore")

<div style="color:#FF0000         ; display  :fill; border-radius:90px;
           background-color:#FFFF00; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :##FF0000; 
          text-align: Left;   font-size:28px; font-family  :cursive">       
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;Importing the Data Set<a id=5></a></p>
</div>

In [2]:
df = pd.read_csv('weatherHistory.csv')
# view dimensions of dataset
print("The Dimensions of Dataset is : ") 
df.shape

The Dimensions of Dataset is : 


(96453, 11)

In [3]:
#Previewing the Data set
df.head()

,Formatted Date,Summary,Precip Type,Temperature (C),Apparent Temperature (C),Humidity,Wind Speed (km/h),Wind Bearing (degrees),Visibility (km),Pressure (millibars),Daily Summary
0,2006-04-01 00:00:00.000 +0200,Partly Cloudy,rain,9.472222,7.388889,0.89,14.1197,251,15.8263,1015.13,Partly cloudy throughout the day.
1,2006-04-01 01:00:00.000 +0200,Partly Cloudy,rain,9.355556,7.227778,0.86,14.2646,259,15.8263,1015.63,Partly cloudy throughout the day.
2,2006-04-01 02:00:00.000 +0200,Mostly Cloudy,rain,9.377778,9.377778,0.89,3.9284,204,14.9569,1015.94,Partly cloudy throughout the day.
3,2006-04-01 03:00:00.000 +0200,Partly Cloudy,rain,8.288889,5.944444,0.83,14.1036,269,15.8263,1016.41,Partly cloudy throughout the day.
4,2006-04-01 04:00:00.000 +0200,Mostly Cloudy,rain,8.755556,6.977778,0.83,11.0446,259,15.8263,1016.51,Partly cloudy throughout the day.


In [4]:
df.tail()

,Formatted Date,Summary,Precip Type,Temperature (C),Apparent Temperature (C),Humidity,Wind Speed (km/h),Wind Bearing (degrees),Visibility (km),Pressure (millibars),Daily Summary
96448,2016-09-09 19:00:00.000 +0200,Partly Cloudy,rain,26.016667,26.016667,0.43,10.9963,31,16.1000,1014.36,Partly cloudy starting in the morning.
96449,2016-09-09 20:00:00.000 +0200,Partly Cloudy,rain,24.583333,24.583333,0.48,10.0947,20,15.5526,1015.16,Partly cloudy starting in the morning.
96450,2016-09-09 21:00:00.000 +0200,Partly Cloudy,rain,22.038889,22.038889,0.56,8.9838,30,16.1000,1015.66,Partly cloudy starting in the morning.
96451,2016-09-09 22:00:00.000 +0200,Partly Cloudy,rain,21.522222,21.522222,0.60,10.5294,20,16.1000,1015.95,Partly cloudy starting in the morning.
96452,2016-09-09 23:00:00.000 +0200,Partly Cloudy,rain,20.438889,20.438889,0.61,5.8765,39,15.5204,1016.16,Partly cloudy starting in the morning.


# GETTING STARTED

<div style="color:#FF0000         ; display  :fill; border-radius:90px;
           background-color:#FFFF00; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :##FF0000; 
          text-align: Left;   font-size:28px; font-family  :cursive">       
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;Missing Value Detection<a id=5></a></p>
</div>

In [5]:
# get the number of missing data points per column
missing_value_count = (df.isnull().sum())
print(missing_value_count[missing_value_count > 0])
# percentage of missing data
total_cells = np.prod(df.shape)
total_missing_value = missing_value_count.sum()
print("Total percentage of our missing value is:",round((total_missing_value / total_cells * 100),4))
print('Total number of our cells is :',total_cells)
print('Total number of our missing value is :',total_missing_value)

Precip Type    517
dtype: int64
Total percentage of our missing value is: 0.0487
Total number of our cells is : 1060983
Total number of our missing value is : 517


<div style="color:#FF0000         ; display  :fill; border-radius:90px;
           background-color:#FFFF00; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :##FF0000; 
          text-align: Left;   font-size:28px; font-family  :cursive">       
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;Handling Missing Values<a id=5></a></p>
</div>

In [6]:
df['Precip Type'].value_counts()

rain    85224
snow    10712
Name: Precip Type, dtype: int64

* As we can see that we have 517 missing values and unfortunately we can not **delete it**, because removing 517 index means removing 517 **random Formatted Date** which can have a **large negative impact** on our Analysis.
* As we can see that the majority of Precip Type is rain so that is why we will replace the missing values with rain variable.

In [7]:
# replace the missing values with rain variable
df['Precip Type'] = df['Precip Type'].fillna("rain")

**- So after filling the missing values, let's check it more again:**

In [8]:
df.isnull().sum()

Formatted Date              0
Summary                     0
Precip Type                 0
Temperature (C)             0
Apparent Temperature (C)    0
Humidity                    0
Wind Speed (km/h)           0
Wind Bearing (degrees)      0
Visibility (km)             0
Pressure (millibars)        0
Daily Summary               0
dtype: int64

**- So Now let's delete 'Daily Summary', we do not need it because we do have 'Summary' feature we will use that.**

In [9]:
df.drop(["Daily Summary"], axis=1, inplace=True)

**- Let's check it:**

In [10]:
df.columns

Index(['Formatted Date', 'Summary', 'Precip Type', 'Temperature (C)',
       'Apparent Temperature (C)', 'Humidity', 'Wind Speed (km/h)',
       'Wind Bearing (degrees)', 'Visibility (km)', 'Pressure (millibars)'],
      dtype='object')

# FEATURE ENGINEERING

<div style="color:#FF0000         ; display  :fill; border-radius:90px;
           background-color:#FFFF00; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :##FF0000; 
          text-align: Left;   font-size:28px; font-family  :cursive">       
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;Handling Dates<a id=5></a></p>
</div>

In [11]:
#Changing Formatted Date from String to Datetime
df['Formatted Date'] = pd.to_datetime(df['Formatted Date'],utc=True)
df['Formatted Date'][0]

Timestamp('2006-03-31 22:00:00+0000', tz='UTC')

In [12]:
df = df.sort_values(by='Formatted Date')

<div style="color:#FF0000         ; display  :fill; border-radius:90px;
           background-color:#FFFF00; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :##FF0000; 
          text-align: Left;   font-size:28px; font-family  :cursive">       
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;Creating Monthly And Daily Dataframe<a id=5></a></p>
</div>

In [13]:
# this DataFrame will use Formatted Date as an index just for visualization objective
data = df.copy()
# Set 'Formatted Date' as index and resample Date
resampled_df = data.set_index('Formatted Date')
resampled_days = resampled_df.resample('D').mean(numeric_only=True)
resampled_df = resampled_df.resample('ME').mean(numeric_only=True)

In [14]:
len(resampled_df),len(df),len(resampled_days)

(133, 96453, 4019)

In [15]:
resampled_df.head(1)

,Temperature (C),Apparent Temperature (C),Humidity,Wind Speed (km/h),Wind Bearing (degrees),Visibility (km),Pressure (millibars)
Formatted Date,,,,,,,
2005-12-31 00:00:00+00:00,0.577778,-4.05,0.89,17.1143,140.0,9.982,1016.66


**Resample('ME')** simply converts the hourly data to monthly by taking the mean. But it will also drop all the "object" columns. So what we can do later is concatenate these "object" columns with all kinds of Formatted Date columns in one Data Frame (for a visualization purpose).

And **Resample('D')** does the exact same idea but it converts the time to daily data

Now after changing Formatted Date from String to Datetime we will replace every date column with a set of date metadata columns, such as holiday, day of week, and month. These columns provide categorical data that we suspect will be useful.

fastai comes with a function that will do this for us—we just have to pass a column name that contains dates:

In [16]:
# make a Date copy because add_datepart deletes the original Formatted Date column
# and replaces it with its date parts
df['Date'] = df["Formatted Date"]
df = add_datepart(df, 'Formatted Date')

In [17]:
df[2000:2001]

,Summary,Precip Type,Temperature (C),Apparent Temperature (C),Humidity,Wind Speed (km/h),Wind Bearing (degrees),Visibility (km),Pressure (millibars),Date,...,Formatted Day,Formatted Dayofweek,Formatted Dayofyear,Formatted Is_month_end,Formatted Is_month_start,Formatted Is_quarter_end,Formatted Is_quarter_start,Formatted Is_year_end,Formatted Is_year_start,Formatted Elapsed
5504,Breezy and Partly Cloudy,rain,3.866667,-1.355556,0.7,29.3986,160,9.982,1016.19,2006-03-25 07:00:00+00:00,...,25,5,84,False,False,False,False,False,False,1.143270e+09


#### We can see that there are now lots of new columns in our DataFrame.